File: python/connector/mysql_connector.py
from config.config import db_user, db_pass, db_host, db_port, db_database

import mysql.connector


class MySQLResults:
    def __init__(self, connector, cursor):
        self.connector = connector
        self.cursor = cursor

    def __enter__(self):
        return self.cursor

    def __exit__(self, exception_type, exception_value, traceback):
        self.cursor.close()
        self.connector.close()


class MySQLConnector:
    """MySQLに接続しデータを取得する関数群が定義されているクラス"""

    def fetchModels(self, query: str) -> MySQLResults:
        """SQLクエリを実行しデータ全てを返す関数

        Args:
            query (str): データ取得クエリ (Selectなど) を想定

        Returns:
            MySQLResults: データアクセスのためのクラス。基本的にはwith句と共に使い、for文で内部データを取得できる
        """
        conn = mysql.connector.connect(user=db_user, password=db_pass, host=db_host, port=db_port, database=db_database)
        cursor = conn.cursor()
        cursor.execute(query)

        return MySQLResults(conn, cursor)

    def fetchModelByID(self, query: str, id: int) -> MySQLResults:
        """SQLクエリを実行し指定したIDデータを返す関数

        Args:
            query (str): データ取得クエリ (Selectなど) を想定
            id (int): SQLクエリに渡すID

        Returns:
            MySQLResults: データアクセスのためのクラス。詳しくは `fetchModels` のドキュメント参照
        """
        conn = mysql.connector.connect(user=db_user, password=db_pass, host=db_host, port=db_port, database=db_database)
        cursor = conn.cursor()
        cursor.execute(query, (id,))

        return MySQLResults(conn, cursor)

File: cpp/src/connector/mysql_connector.cpp
#pragma once

#include <cppconn/driver.h>
#include <cppconn/prepared_statement.h>
#include <cppconn/resultset.h>
#include <cppconn/statement.h>
#include <mysql_driver.h>

#include <memory>
#include <mutex>
#include <string>

#include "../config/config.cpp"

namespace geniee::connector::mysql {
    /// @brief
    /// MySQLに接続しデータを取得する関数群が定義されているクラス。使用ライブラリの都合上MySQLの接続はスレッドセーフではないため、こちらのクラスで排他制御。
    class MySQLConnector {
    public:
        /// @brief SQLクエリを実行しデータ全てを返す関数
        /// @param query SQLクエリ。データ取得クエリ (Selectなど) を想定
        /// @return データアクセスのためのクラス。基本的には `next()`　でCursorを動かし `getXXX()` でデータを取得する。詳しくは
        /// `fetch_model_from_db.cpp` か公式リファレンスを参照のこと。sql::ResultSetはdeleteする必要があるがそちらはunique_ptrが面倒を見る。
        std::unique_ptr<sql::ResultSet> fetchModels(std::string query) {
            std::lock_guard<std::mutex> lock(mut);
            sql::mysql::MySQL_Driver *driver = sql::mysql::get_mysql_driver_instance();  // この関数はthread safe *ではないので* 排他制御
            std::unique_ptr<sql::Connection> con(driver->connect(geniee::config::db_uri, geniee::config::db_user, geniee::config::db_pass));
            std::unique_ptr<sql::Statement> stmt(con->createStatement());
            return std::unique_ptr<sql::ResultSet>(stmt->executeQuery(query));
            /* NOTE: But do not explicitly free driver, the connector object. Connector/C++ takes care of freeing that.
             * driverを呼び出し側がdeleteしてはいけない。mysql側が面倒を見ている。
             * https://dev.mysql.com/doc/connector-cpp/1.1/en/connector-cpp-examples-connecting.html
             */
        }

        /// @brief SQLクエリを実行し指定したIDデータを返す関数。
        /// @param query SQLクエリ。データ取得クエリ (Selectなど) を想定
        /// @param id SQLクエリに渡すID。
        /// @return データアクセスのためのクラス。使い方は `geniee::connector::mysql::fetchModels` のドキュメント参照。
        std::unique_ptr<sql::ResultSet> fetchModelByID(std::string query, unsigned int id) {
            std::lock_guard<std::mutex> lock(mut);
            sql::mysql::MySQL_Driver *driver = sql::mysql::get_mysql_driver_instance();  // この関数はthread safe *ではないので* 排他制御
            std::unique_ptr<sql::Connection> con(driver->connect(geniee::config::db_uri, geniee::config::db_user, geniee::config::db_pass));
            std::unique_ptr<sql::PreparedStatement> stmt(con->prepareStatement(query));
            stmt->setUInt(1, id);  // ここでIDをセット。
            return std::unique_ptr<sql::ResultSet>(stmt->executeQuery());
        }

    private:
        std::mutex mut;
    };
}  // namespace geniee::connector::mysql
